### ISSUE TITLE
[RFC]: Introduce ATOM as model implementation backend for AMD GPU

### ISSUE THREAD

**Issue by zejunchen-zejun** [vllm-project/vllm#33478]
### Motivation
[ATOM](https://github.com/ROCm/ATOM) is a foundational component of AMD’s AI inference strategy. It is implemented by ROCm developers and can be used to serve as the model implementation backend for high-performance inference on AMD GPUs. It is built by integrating optimizations from ROCm’s high-performance operator library [aiter](https://github.com/ROCm/aiter/tree/main) and high-performance communication library [mori](https://github.com/ROCm/mori) into model execution path.

The key motivation of introducing ATOM into vLLM as a model impl backend is two-fold:
- Strengthen vLLM’s ecosystem for AMD GPUs. ATOM, working as the ROCm-optimized model impl backend, can help vLLM deliver more competitive performance on AMD GPUs
- Balance user experience of AMD GPUs and ROCm iteration velocity:
  - A great out-of-the-box experience for vLLM frontend users (only one serving argument `--model-impl atom` to enable, reuse vLLM’s serving/runtime stack)
  - Fast, continuous development iteration on ROCm libraries (aiter/mori) and their model-side integration

### Background
ATOM can deliver performance gains through following points with ROCm components:
- Cross-layer / cross-module fusion opportunities (spanning layer boundaries)
- Specific KV Cache layout required by kernel
- New optimized ops and kernel implementations tailored for AMD GPU characteristics
- Communication and runtime optimizations

While ATOM primarily focuses on model-level and kernel-level optimizations, it is designed to integrate with and leverage the high-level framework features. Meanwhile vLLM is most popular and successful framework, where many features are developed and many hardware devices are supported. Additionally, vLLM is very open to cooperate with the products from the community, for example the [transformers](https://github.com/huggingface/transformers) from huggingface and the [terratorch](https://github.com/terrastackai/terratorch) from IBM. Both of them have already successfully used the model implementation backend registry of vLLM and benefit many users. This feature has been proven to be used for:
- Swapping model execution implementations while continuing to benefit from vLLM’s existed excellent design, like scheduler, kv cache manager, cuda graph compilation, parallel strategies, PD infra, profiler and so on
- Preserving the vLLM stack and its user experience

Given above points, ATOM plans to integrate into vLLM as a model impl backend. This integration aligns with vLLM’s extensible architecture—leveraging the existing model implementation backend registry without requiring changes to core components. By building on this well-designed extension point, ATOM can deliver hardware-aware model-level optimizations while fully respecting vLLM’s separation of concerns and commitment to maintainability. Given vLLM’s position as the de facto standard for high-performance LLM serving, this approach ensures that AMD GPU users can benefit from the latest model-level optimizations without fragmenting the ecosystem or compromising the consistency that vLLM users rely on.

### Design Overview
Here is a diagram showing how ATOM works as the model impl backend with vLLM.

<img width="975" height="708" alt="Image" src="https://github.com/user-attachments/assets/87e896fc-c3d0-4e38-84ad-d086b33e6481" />

When frontend users launch vLLM with `--model-impl atom`, vLLM will instantiate a thin wrapper model that delegates the heavy lifting to ATOM:
- vLLM creates `ATOMForCausalLM` / `ATOMMoEForCausalLM` for dense / sparse model. Both of them inherits from the `ATOMModelBase`.
- The wrapper calls into ATOM’s plugin entrypoint when users have installed the ATOM. The entrypoint is shown as follow:
  - `model = atom.prepare_model(config=vllm_config, engine="vllm")`
- ATOM constructed an optimized model object (a callable) that implements the construction / forward / weight loading / logits computation methods, aligned with vLLM calling convention:
   - `__init__(...)`
   - `forward(...)`
   - `load_weights(...)`
   - `compute_logits(...)`

This is intentionally aligned with vLLM, so that the ATOM backend owns model-specific kernels and internal execution details.

### Attention
The vLLM’s attention backend selection (including a `CUSTOM` backend placeholder) is a particularly clean design for enabling the alternative custom attention for preserving the vLLM’s high-level attention design. ATOM leverages this registry mechanism and provide the required attention implementation class and metadata builder. Here is the associated code for registering atom attention backend to vLLM.
```
register_backend(backend=AttentionBackendEnum.CUSTOM,
                 is_mamba=False,
                 class_path="atom.model_ops.attentions.aiter_attention.AiterBackend")
```
In practice, the frontend users can run with the environment variable `VLLM_ATTENTION_BACKEND=CUSTOM`

### Execution Flow of vLLM with ATOM backend
Here is the execution flow of running workloads on vLLM with ATOM model impl backend
<img width="717" height="907" alt="Image" src="https://github.com/user-attachments/assets/6c8be829-1be2-4a01-8a0a-ba8b45cb4bbf" />

### Current Status
Here is the performance value of vLLM and vLLM with ATOM model impl backend for Qwen3-235B-A22B-Instruct-2507-FP8 with TP8+EP8:
Machine   350*8 | ISL/OSL     (K=1000) | Concurrency | TTFT(ms) | TPOT(ms) | TTPS(tok/s) | TTPS Ratio     (Compared to vLLM)
-- | -- | -- | -- | -- | -- | --
vLLM | ISL  : 3.0-3.6K     OSL: 0.3-0.5K | 220 | 273.6406 | 53.1735 | 32527.4533 | 100.0%
vLLM(ATOM   Model Impl Backend) | ISL  : 3.0-3.6K     OSL: 0.3-0.5K | 220 | 281.4742 | 51.8225 | 33444.0514 | 102.8%
vLLM | ISL  : 0.8-1K     OSL: 1.6-2K | 512 | 217.1111 | 46.9111 | 15104.8034 | 100.0%
vLLM(ATOM   Model Impl Backend) | ISL  : 0.8-1K     OSL: 1.6-2K | 512 | 207.0059 | 44.0972 | 16083.9545 | 106.5%
vLLM | ISL  : 3.6-4.4K     OSL: 1.8-2.2K | 260 | 266.9704 | 46.6293 | 15520.3933 | 100.0%
vLLM(ATOM   Model Impl Backend) | ISL  : 3.6-4.4K     OSL: 1.8-2.2K | 260 | 275.815 | 43.4682 | 16596.3231 | 106.9%

Here is the accuracy value of vLLM with ATOM model impl backend:
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     3|exact_match|↑  |0.8901|±  |0.0086|
|     |       |strict-match    |     3|exact_match|↑  |0.8772|±  |0.0090|

### PRs
https://github.com/vllm-project/vllm/pull/32160

**Comment by sunway513** [comment]
Thanks for the thoughtful feedback, @simon-mo and @ProExpertProg. We agree with the goal of converging on a single, canonical vLLM model implementation, and see ATOM primarily as a vehicle to rapidly develop, validate, and upstream ROCm-specific operator and runtime improvements.

Our intent is to continue prioritizing in-tree native operator integration in vLLM (e.g., AITER, MORI, attention and sampling kernels), along with sustained upstream contributions. Based on the discussion, we will reframe the RFC to focus on a plugin-oriented approach aligned with vLLM’s existing extension points, while upstreaming proven pieces into the core over time.

We are committed to staying closely engaged with the vLLM community and iterating quickly in the open to ensure AMD platforms are first-class citizens in the canonical vLLM path.

**Comment by zejunchen-zejun** [comment]
With the constructive engagement and consensus reached between the AMD team and the vLLM community, this RFC will be formally closed. Thank you all.

**Comment by hmellor** [comment]
> The transformers backend exists specifically as a catch-all for models not yet supported natively

As a side note, the Transformers backend does actually use as many vLLM native modules as it can, rather than being a black box. So it _is actually_ using vLLM's `Attention` and `FusedMoE` for example.
